velox/exec/MemoryReclaimer.cpp

```diff
@@ -66,6 +66,106 @@ void MemoryReclaimer::abort(
   });
 }
 
+/*static*/ std::unique_ptr<memory::MemoryReclaimer> NodeMemoryReclaimer::create(
+    folly::Executor* executor) {
+  return std::unique_ptr<memory::MemoryReclaimer>(
+      new NodeMemoryReclaimer(executor));
+}
+
+NodeMemoryReclaimer::NodeMemoryReclaimer(folly::Executor* executor)
+    : executor_(executor) {}
+
+uint64_t NodeMemoryReclaimer::reclaim(
+    memory::MemoryPool* pool,
+    uint64_t targetBytes,
+    uint64_t maxWaitMs,
+    Stats& stats) {
+  if (executor_ == nullptr) {
+    return memory::MemoryReclaimer::reclaim(
+        pool, targetBytes, maxWaitMs, stats);
+  }
+
+  // Sort the child pools based on their reserved memory and reclaim from the
+  // child pool with most reservation first.
+  struct Candidate {
+    std::shared_ptr<memory::MemoryPool> pool;
+    uint64_t reclaimableBytes;
+  };
+  std::vector<Candidate> candidates;
+  {
+    std::shared_lock guard{pool->poolMutex_};
+    candidates.reserve(pool->children_.size());
+    for (auto& entry : pool->children_) {
+      auto child = entry.second.lock();
+      if (child != nullptr) {
+        const uint64_t reclaimableBytes = child->reclaimableBytes().value_or(0);
+        candidates.push_back(Candidate{std::move(child), reclaimableBytes});
+      }
+    }
+  }
+  struct ReclaimResult {
```

Comment:
Candidate and ReclaimResult can be put in the base class. I remember there were other child impl uses similar structures as well?

Reply:
They are different types.